Deprecate remaining VI methods

purna135 opened the pull request:
Addressing #5581

Replaced remaining NotImplementedInference methods with deprecation warning.

purna135 committed:
Deprecated remaining VI methods

purna135 commented:
I need some help with the next step; is adding a deprecation warning the right way to go?

ghost commented:
I think the idea was to just remove them (even though the title in the issue says "deprecate"). The methods are not implemented in v4, so there is really nothing to deprecate.

purna135 commented:
Could someone please help me remove the Normalizing flows?
I'm not sure if I should remove the entire NormalizingFlowGroup class or just a few lines of code.

https://github.com/pymc-devs/pymc/blob/ef5f91b41d733c64cf3ea785ec2760caae5d1bd4/pymc/variational/approximations.py#L362-L364

canyon289 commented:
@ferrine would you like to take a look here?

ferrine commented on pymc/variational/approximations.py:
```diff
@@ -270,14 +272,14 @@ def _new_initial(self, size, deterministic, more_replacements=None):
             )
+        elif deterministic:
+            warnings.warn(
+                "Deterministic sampling from a Histogram is deprecated and will be removed soon.",
```
Is it indeed broken or it was a false alarm?

ferrine replied:
We can enable the test for that and check

ferrine commented on pymc/variational/opvi.py:
```diff
@@ -957,9 +957,9 @@ def __init_group__(self, group):
         if not group:
             raise GroupError("Got empty group")
         if self.local:
```
I think the context of the issue was to entirely delete the code and `local` kwarg, feel free to just delete large portions of the code.

ferrine commented on pymc/variational/opvi.py:
```diff
@@ -957,9 +957,9 @@ def __init_group__(self, group):
         if self.local:
-            raise NotImplementedInference("Local inferene aka AEVB is not supported in v4")
+            warnings.warn("Local inferene aka AEVB is deprecated.", DeprecationWarning)
         if self.batched:
```
Same for `batched` kwarg and all its mentions. Just delete without hesitation

ferrine reviewed (commented):
As for normalizing flows I expect something like `rm pymc/variational/flows.py` as a good first step

fonnesbeck commented:
Is this enough to go on @purna135 ?

purna135 commented:
I'll work on these suggestions and let you know if I need anything else.
Thank you very much.

michaelosthege commented:
Hi @purna135, this PR is one of the final steps towards `v4.0.0`. Could you pick it up again and push it over the finish line?

purna135 commented:
Hi @michaelosthege sorry for the delay; I was preoccupied with my exam, which has now been done.
I will complete this RP in few days.

twiecki commented on pymc/variational/opvi.py:
```diff
@@ -957,9 +957,9 @@ def __init_group__(self, group):
             raise GroupError("Got empty group")
         if self.local:
-            raise NotImplementedInference("Local inferene aka AEVB is not supported in v4")
+            warnings.warn("Local inferene aka AEVB is deprecated.", DeprecationWarning)
```
We should just delete this code, rather than deprecate.

purna135 committed:
Deleted flows and code for local and batched kwarg

purna135 commented:
Hi @twiecki and @ferrine could you please have a look if this is the right way to go?

ferrine reviewed (approved):
Looks good to me, can be merged once tests pass

purna135 commented:
Thank you very much, @ferrine. Could you kindly advise me on how to pass these failed tests?
In the doc string, I realised I left some references to `batched` and `local`, which I will remove in my next commit.

ferrine commented:
@purna135, the tests diff was collapsed. I think I missed something here. I have to go through it once more

ferrine commented on pymc/tests/test_variational_inference.py:
```diff
@@ -98,89 +87,6 @@ def three_var_model():
-        (not_raises(), {MeanFieldGroup: ["one"], FullRankGroup: ["two", "three"]}),
-    ],
-)
-@ignore_not_implemented_inference
```
the `ignore_not_implemented_inference` hook was used to ignore tests that fail due to the not supported functionality. 

`NotImplementedInference` is an explicit error not to confuse failures with other "bugs". Most of the tests that are deleted so far are valid tests. 

The commit that deletes all these tests should be reverted. Or, if not easy, you can use the original file that is present in master.

To summarise. 

1. The only obsolete thing in the test file is `ignore_not_implemented_inference`. 
2. Once it is removed, all tests should pass and no `NotImplementedInference` is raised.
3. The only tests to remove are that fail because of the removed functionality.

ferrine reviewed (changes requested):
I missed the collapsed diff of tests. Most of the deleted tests are valid. I've left a comment above with the explanation.

ferrine commented:
We did a quick video chat with @purna135 to avoid further confusion. Additional thing to revisit was the documentation. That missed my eye as well.

purna135 commented:
Thank you so much for your time and help @ferrine : )

purna135 committed:
Merge remote-tracking branch 'upstream/main' into deprecate_vi_methods

purna135 committed:
removed rowwise from opvi.py

purna135 committed:
redo the test_variational_inference.py

purna135 committed:
refactored some test cases

purna135 committed:
removed mention of local from docstring

purna135 committed:
resolved merge conflict

purna135 commented:
Hello, @ferrine. I deleted several flow-related tests and refactored others, but there are still 10 tests that I am unable to refactor and have commented so that other tests can be checked. Could you please assist me in refactoring the tests I commented? 
Many thanks

purna135 committed:
corrected 'MeanFieldGroup' object has no attribute 'rng'
